src/syntax/ast.rs

```diff
@@ -5,101 +5,122 @@ pub struct Resource {
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Comment {
-    pub body: String,
-}
-
-#[derive(Debug, PartialEq)]
-pub struct Section {
-    pub key: Key,
-    pub body: Vec<Entry>,
+pub struct Message {
+    pub id: Identifier,
+    pub value: Option<Pattern>,
+    pub attributes: Option<Vec<Attribute>>,
     pub comment: Option<Comment>,
 }
 
 #[derive(Debug, PartialEq)]
 pub enum Entry {
     Message(Message),
-    Comment(Comment),
     Section(Section),
+    Comment(Comment),
     Junk(JunkEntry),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Message {
+pub struct Pattern {
+    pub elements: Vec<PatternElement>,
+    pub quoted: bool,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Expression {
+    Selector(SelectorExpression),
+    SelectExpression {
+        exp: Option<SelectorExpression>,
+        variants: Vec<Variant>
+    },
+}
+
+#[derive(Debug, PartialEq)]
+pub enum SelectorExpression {
+    Pattern(Pattern),
+    MessageReference(String),
+    ExternalArgument(String),
+    CallExpression {
+        callee: Builtin,
+        args: Vec<SelectorExpression>,
+    },
+    VariantExpression {
+        id: Identifier,
+        key: VariantKey
+    },
+    AttributeExpression {
+        id: Identifier,
+        attr: Identifier
```

Comment:
The ASDL uses `msg` and `id` here. I like your idea of using `id` instead of `msg`.  I would suggest `name` instead of `attr`. `attr` suggest it's the whole attribute object, just like `var` would suggest it's the whole variant and not a variant key in `VariantExpression`.